File: python/src/loadstrike_sample_reference/methods/loadstrike_scenario/with_clean_async_method_reference.py
from __future__ import annotations

from typing import Any

from loadstrike_sdk import (
    CorrelationStoreConfiguration,
    CrossPlatformScenarioConfigurator,
    HttpEndpointDefinition,
    LoadStrikeMetric,
    LoadStrikePluginData,
    LoadStrikePluginDataTable,
    LoadStrikeReportFormat,
    LoadStrikeResponse,
    LoadStrikeRunner,
    LoadStrikeScenario,
    LoadStrikeSimulation,
    LoadStrikeStep,
    LoadStrikeThreshold,
    RedisCorrelationStoreOptions,
    TrackingFieldSelector,
    TrackingPayloadBuilder,
)

RUNNER_KEY = "runner_dummy_orders_reference"

class TempConfigPaths:
    def __init__(self, config_path: str, infra_path: str) -> None:
        self.config_path = config_path
        self.infra_path = infra_path


class OrdersReportingSink:
    @property
    def sink_name(self) -> str:
        return "orders-sample-sink"

    def init(self, _context: Any, _infra_config: Any) -> None:
        return

    def start(self, _session: Any) -> None:
        return

    def save_realtime_stats(self, _stats: Any) -> None:
        return

    def save_realtime_metrics(self, _metrics: Any) -> None:
        return

    def save_run_result(self, _result: Any) -> None:
        return

    def stop(self) -> None:
        return


class OrdersRuntimePolicy:
    def should_run_scenario(self, _scenario_name: str) -> bool:
        return True

    def before_scenario(self, _scenario_name: str) -> None:
        return

    def after_scenario(self, _scenario_name: str, _stats: Any) -> None:
        return

    def before_step(self, _scenario_name: str, _step_name: str) -> None:
        return

    def after_step(self, _scenario_name: str, _step_name: str, _reply: Any) -> None:
        return


class OrdersWorkerPlugin:
    plugin_name = "orders-sample-plugin"

    def init(self, _context: Any = None, _infra_config: Any = None) -> None:
        return

    def start(self, _session: Any = None) -> None:
        return

    def get_data(self, _result: Any) -> LoadStrikePluginData:
        return LoadStrikePluginData.create(self.plugin_name)

    def stop(self) -> None:
        return

    def dispose(self) -> None:
        return


def create_order_reply():
    return LoadStrikeResponse.ok("200", 128, "ok", 3.2)


def execute_order_get(context: Any):
    return LoadStrikeStep.run("get-order", context, lambda: create_order_reply()).as_reply()


def baseline_scenario(name: str = "orders.get-by-id"):
    return (
        LoadStrikeScenario.create(name, execute_order_get)
        .with_load_simulations(LoadStrikeSimulation.iterations_for_constant(1, 1))
        .without_warm_up()
    )


def base_runner():
    return (
        LoadStrikeRunner.create()
        .add_scenario(baseline_scenario())
        .with_runner_key(RUNNER_KEY)
        .with_test_suite("orders-reference")
        .with_test_name("orders-get-by-id")
        .without_reports()
    )


def base_context():
    return base_runner().build_context()


def http_source():
    return HttpEndpointDefinition(
        name="orders-http-source",
        mode="Produce",
        tracking_field=TrackingFieldSelector.parse("header:X-Correlation-Id"),
        url="https://orders.example.test/api/orders",
        method="GET",
        response_source="Body",
    )


def http_destination():
    return HttpEndpointDefinition(
        name="orders-http-destination",
        mode="Consume",
        tracking_field=TrackingFieldSelector.parse("json:$.trackingId"),
        gather_by_field=TrackingFieldSelector.parse("json:$.tenantId"),
        url="https://orders.example.test/api/order-events",
        method="GET",
        response_source="Body",
        consume_json_array_response=True,
        consume_array_path="$.items",
    )


def tracking_configuration():
    return {
        "Source": http_source(),
        "Destination": http_destination(),
        "RunMode": "GenerateAndCorrelate",
        "CorrelationTimeoutSeconds": 30,
        "TimeoutSweepIntervalSeconds": 1,
        "TimeoutBatchSize": 200,
        "TimeoutCountsAsFailure": True,
        "MetricPrefix": "orders_tracking",
        "ExecuteOriginalScenarioRun": False,
        "CorrelationStore": CorrelationStoreConfiguration.in_memory(),
    }


def tracked_context():
    return (
        LoadStrikeRunner.create()
        .add_scenario(
            CrossPlatformScenarioConfigurator.configure(
                baseline_scenario("orders.tracked"),
                tracking_configuration(),
            )
        )
        .with_runner_key(RUNNER_KEY)
        .without_reports()
        .build_context()
    )


def build_tracking_payload():
    builder = TrackingPayloadBuilder()
    builder.set_body('{"trackingId":"ord-1"}')
    return builder.build()


def run_result():
    return None


def scenario_stats():
    return None


def write_temp_config_files():
    return TempConfigPaths(
        "method-reference.loadstrike.config.json",
        "method-reference.loadstrike.infra.json",
    )

class WithCleanAsyncMethodReference:
    @staticmethod
    def attach_async_no_op_hook_example() -> Any:
        """Attach an asynchronous lifecycle hook with no extra side effects."""
        return baseline_scenario().with_clean_async(lambda context: None)

    @staticmethod
    def attach_async_metric_hook_example() -> Any:
        """Attach an asynchronous lifecycle hook that still registers a metric."""
        return baseline_scenario().with_clean_async(lambda context: context.register_metric(LoadStrikeMetric.counter("orders_seen", "count")))

File: go/src/methods/loadstrike_scenario/with_clean_async_method_reference.go
package loadstrike_scenario

        import (
        	"time"

        	loadstrike "loadstrike.com/sdk/go"
        )

        const withCleanAsyncRunnerKey = "runner_dummy_orders_reference"

type WithCleanAsyncMethodReference struct{}

type withCleanAsyncTempConfigPaths struct {
	ConfigPath string
	InfraPath  string
}

type withCleanAsyncOrdersReportingSink struct{}

func newWithCleanAsyncOrdersReportingSink() withCleanAsyncOrdersReportingSink {
	return withCleanAsyncOrdersReportingSink{}
}
func (withCleanAsyncOrdersReportingSink) SinkName() string {
	return "orders-sample-sink"
}
func (withCleanAsyncOrdersReportingSink) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersReportingSink) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersReportingSink) SaveRealtimeStats([]loadstrike.LoadStrikeScenarioStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersReportingSink) SaveRealtimeMetrics(loadstrike.LoadStrikeMetricStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersReportingSink) SaveRunResult(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersReportingSink) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersReportingSink) Dispose() {}

type withCleanAsyncOrdersRuntimePolicy struct{}

func newWithCleanAsyncOrdersRuntimePolicy() withCleanAsyncOrdersRuntimePolicy {
	return withCleanAsyncOrdersRuntimePolicy{}
}
func (withCleanAsyncOrdersRuntimePolicy) ShouldRunScenario(string) loadstrike.LoadStrikeBoolTask {
	return loadstrike.TaskFromBool(true)
}
func (withCleanAsyncOrdersRuntimePolicy) BeforeScenario(string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersRuntimePolicy) AfterScenario(string, loadstrike.LoadStrikeScenarioRuntime) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersRuntimePolicy) BeforeStep(string, string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersRuntimePolicy) AfterStep(string, string, loadstrike.LoadStrikeReply) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

type withCleanAsyncOrdersWorkerPlugin struct{}

func newWithCleanAsyncOrdersWorkerPlugin() withCleanAsyncOrdersWorkerPlugin {
	return withCleanAsyncOrdersWorkerPlugin{}
}
func (withCleanAsyncOrdersWorkerPlugin) PluginName() string {
	return "orders-sample-plugin"
}
func (withCleanAsyncOrdersWorkerPlugin) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersWorkerPlugin) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersWorkerPlugin) GetData(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikePluginData] {
	return loadstrike.TaskFromResult(loadstrike.CreatePluginData("orders-sample-plugin"))
}
func (withCleanAsyncOrdersWorkerPlugin) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withCleanAsyncOrdersWorkerPlugin) Dispose() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

func withCleanAsyncPerformOrderGetReply() loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeResponse.Ok("200", int64(128), "ok", loadstrike.TimeSpan(3*time.Millisecond))
}

func withCleanAsyncExecuteOrderGet(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeStep.Run("get-order", context, func(loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
		return withCleanAsyncPerformOrderGetReply()
	})
}

func withCleanAsyncExecuteOrderGetAsync(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikeReply] {
	return loadstrike.TaskFromResult(withCleanAsyncExecuteOrderGet(context))
}

func withCleanAsyncBaselineScenario(names ...string) loadstrike.LoadStrikeScenario {
	name := "orders.get-by-id"
	if len(names) > 0 && names[0] != "" {
		name = names[0]
	}
	return loadstrike.CreateScenario(name, withCleanAsyncExecuteOrderGet).
		WithLoadSimulations(loadstrike.LoadStrikeSimulation.IterationsForConstant(1, 1)).
		WithoutWarmUp()
}

func withCleanAsyncBaseRunner() loadstrike.LoadStrikeRunner {
	return loadstrike.Create().
		AddScenario(withCleanAsyncBaselineScenario()).
		WithRunnerKey(withCleanAsyncRunnerKey).
		WithTestSuite("orders-reference").
		WithTestName("orders-get-by-id").
		WithoutReports()
}

func withCleanAsyncBaseContext() loadstrike.LoadStrikeContext {
	return withCleanAsyncBaseRunner().BuildContext()
}

func withCleanAsyncHttpSource() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-source",
		Mode:          "Produce",
		TrackingField: "header:X-Correlation-Id",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                   "https://orders.example.test/api/orders",
			Method:                "GET",
			TrackingPayloadSource: "Request",
			ResponseSource:        "ResponseBody",
		},
	}
}

func withCleanAsyncHttpDestination() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-destination",
		Mode:          "Consume",
		TrackingField: "json:$.trackingId",
		GatherByField: "json:$.tenantId",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                      "https://orders.example.test/api/order-events",
			Method:                   "GET",
			ResponseSource:           "ResponseBody",
			ConsumeJSONArrayResponse: true,
			ConsumeArrayPath:         "$.items",
		},
	}
}

func withCleanAsyncTrackingConfiguration() *loadstrike.TrackingConfigurationSpec {
	return &loadstrike.TrackingConfigurationSpec{
		Source:                      withCleanAsyncHttpSource(),
		Destination:                 withCleanAsyncHttpDestination(),
		RunMode:                     "GenerateAndCorrelate",
		CorrelationTimeoutSeconds:   30,
		TimeoutSweepIntervalSeconds: 1,
		TimeoutBatchSize:            200,
		TimeoutCountsAsFailure:      true,
		MetricPrefix:                "orders_tracking",
		ExecuteOriginalScenarioRun:  false,
	}
}

func withCleanAsyncTrackedContext() loadstrike.LoadStrikeContext {
	return loadstrike.Create().
		AddScenario(withCleanAsyncBaselineScenario("orders.tracked").WithCrossPlatformTracking(withCleanAsyncTrackingConfiguration())).
		WithRunnerKey(withCleanAsyncRunnerKey).
		WithoutReports().
		BuildContext()
}

func withCleanAsyncBuildTrackingPayload() loadstrike.TrackingPayload {
	builder := loadstrike.TrackingPayloadBuilder{
		Headers:            map[string]string{"X-Correlation-Id": "ord-1"},
		ContentType:        "application/json",
		MessagePayloadType: "json",
	}
	builder.SetBody(map[string]any{"trackingId": "ord-1"})
	return builder.Build()
}

func withCleanAsyncRunResult() loadstrike.LoadStrikeRunResult {
	var result loadstrike.LoadStrikeRunResult
	return result
}

func withCleanAsyncScenarioStats() loadstrike.LoadStrikeScenarioStats {
	var stats loadstrike.LoadStrikeScenarioStats
	return stats
}

func withCleanAsyncWriteTempConfigFiles() withCleanAsyncTempConfigPaths {
	return withCleanAsyncTempConfigPaths{
		ConfigPath: "method-reference.loadstrike.config.json",
		InfraPath:  "method-reference.loadstrike.infra.json",
	}
}

        // Attach an asynchronous lifecycle hook with no extra side effects.
func (reference WithCleanAsyncMethodReference) AttachAsyncNoOpHookExample() any {
    return withCleanAsyncBaselineScenario().WithCleanAsync(func(context loadstrike.LoadStrikeScenarioInitContext) loadstrike.LoadStrikeTask { return loadstrike.CompletedTask() })
}

// Attach an asynchronous lifecycle hook that still registers a metric.
func (reference WithCleanAsyncMethodReference) AttachAsyncMetricHookExample() any {
    return withCleanAsyncBaselineScenario().WithCleanAsync(func(context loadstrike.LoadStrikeScenarioInitContext) loadstrike.LoadStrikeTask { context.RegisterMetric(loadstrike.Metric.CreateCounter("orders_seen", "count")); return loadstrike.CompletedTask() })
}
